### PR TITLE
RFC: Add Exclude Flag for Targeted Plan/Apply

### PR DESCRIPTION
**Pull request by JamesBalazs**
Links to #426 

## Target Release

1.8.0

## Checklist

- [x] I have read the [contribution guide](https://github.com/opentofu/opentofu/blob/main/CONTRIBUTING.md).
- [x] I have not used an AI coding assistant to create this PR.
- [x] I have written all code in this PR myself OR I have marked all code I have not written myself (including modified code, e.g. copied from other places and then modified) with a comment indicating where it came from.

### Go checklist

 

- [ ] I have run golangci-lint on my change and receive no errors relevant to my code.
- [ ] I have run existing tests to ensure my code doesn't break anything.
- [ ] I have added tests for all relevant use cases of my code, and those tests are passing.
- [ ] I have only exported functions, variables and structs that should be used from other packages.
- [ ] I have added meaningful comments to all exported functions, variables, and structs.

### Website/documentation checklist

- [ ] I have locally started the website as [described here](https://github.com/opentofu/opentofu/blob/main/website/README.md) and checked my changes.

### CODE REVIEW

File: rfc/20240611-ExcludeFlag.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+
+## Proposed Solution
+
+Add a new flag, `-exclude`, to plan and apply. This should allow a plan to be created, or apply to be run, for all resources except those that are "excluded" and their children.
+
+### User Documentation
+
+- Usage example:
+`opentf plan -exclude module.something_broken` / `opentf apply -exclude module.something_broken` should be added to --help output for plan/apply.
```
Comment (ollevche) [inline rfc/20240611-ExcludeFlag.md:19]:
```suggestion
`tofu plan -exclude module.something_broken` / `tofu apply -exclude module.something_broken` should be added to --help output for plan/apply.
```

File: rfc/20240611-ExcludeFlag.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+- `-target` is *not* documented in website/docs/cli/commands/apply.mdx, but probably should be. It would be nice to add documentation for both `-target` and `-exclude` there
+
+### Technical Approach
+
+Technical summary, easy to understand by someone unfamiliar with the codebase.
+
+- `-exclude` flag added to extendedFlagSet when an operation is provided (same as `-target`)
+- Passed through to `OperationRequest`
+- `TargetsTransformer.Transform` should remove vertices that are in (or children of items in) `t.Excludes` if `t.Excludes` are provided (just like it currently removes vertices that _aren't_ in `t.Targets`).
+- If both targets and excludes are provided, use the most specific path eg:
```
Comment (ollevche) [inline rfc/20240611-ExcludeFlag.md:31]: I believe we should postpone this use case (target + exclude). We could also add it later if requested by the community. See https://github.com/opentofu/opentofu/issues/426#issuecomment-2137356757

**Review by ollevche — CHANGES_REQUESTED** [review]
Thank you for the RFC!

File: rfc/20240611-ExcludeFlag.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+Issue: https://github.com/opentofu/opentofu/issues/426
+
+This RFC details a new flag, `-exclude`, which would do the inverse of `-target` - that is, plan/apply all resources except those specified by the exclude flag and their descendants.
+
+Only for use in exceptional circumstances as is the case for `-target`, this makes it a lot easier to plan/apply all changes to an environment except ones with known issues eg where there is a problem with the infrastructure or the committed config for one module/resource, but you still want to apply everything else.
+
+https://github.com/hashicorp/terraform/issues/2253 is one of the most requested features in the Terraform repo
+
+
+## Proposed Solution
```
Comment (RLRabinowitz) [inline rfc/20240611-ExcludeFlag.md:12]: This section is missing the specifics of how the `-exclude` flag will work. For example - what happens to dependencies and dependents of an excluded resource, what happens when there's a mixture of `-target` and `-exclude` flag in the same execution (should not be allowed for now)

For example - following up on [this comment](https://github.com/opentofu/opentofu/issues/426#issuecomment-2137148391) from the issue

File: rfc/20240611-ExcludeFlag.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+- `-target` is documented in website/docs/cli/commands/plan.mdx, so we should also document `-exclude` there
+- `-target` is *not* documented in website/docs/cli/commands/apply.mdx, but probably should be. It would be nice to add documentation for both `-target` and `-exclude` there
+
+### Technical Approach
+
+Technical summary, easy to understand by someone unfamiliar with the codebase.
+
+- `-exclude` flag added to extendedFlagSet when an operation is provided (same as `-target`)
+- Passed through to `OperationRequest`
+- `TargetsTransformer.Transform` should remove vertices that are in (or children of items in) `t.Excludes` if `t.Excludes` are provided (just like it currently removes vertices that _aren't_ in `t.Targets`).
```
Comment (RLRabinowitz) [inline rfc/20240611-ExcludeFlag.md:30]:
```suggestion
- `TargetsTransformer.Transform` should remove vertices that are in `t.Excludes`, or descendants of items in `t.Excludes`, if `t.Excludes` are provided (just like it currently removes vertices that _aren't_ in `t.Targets`).
```

File: rfc/20240611-ExcludeFlag.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+- If both targets and excludes are provided, use the most specific path eg:
+```-exclude module.entire_environment \
+-include module.entire_environment.module.network \
+-include module.entire_environment.module.application \
+-exclude module.entire_environment.module.application.module.eks
+```
+should exclude the whole entire_environment module but include the network and application modules within, except the eks module within the application module. This should be the outcome regardless of the order of the flags.
+- Return an error when `-target` and `-exclude` point at the exact same resource
+- When destroying, if there are dependencies of an excluded resource, the dependencies should also be kept (excluded from change)
```
Comment (RLRabinowitz) [inline rfc/20240611-ExcludeFlag.md:39]: I feel like this statement should be clarified. When destroying, the dependencies are the ones excluded, and not the dependent nodes. This, however, is talking about the non-inverted graph, not to be confused with the inverted graph of the destroy action.

File: rfc/20240611-ExcludeFlag.md [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+https://github.com/opentofu/opentofu/pull/427 this rough proof of concept was previously drafted although there are a few issues
+- not code complete
+- outdated with merge conflicts
+- some unnecessary changes were included
+I would likely start from scratch on the latest main branch, and only include the necessary changes.
+
+Potential limitations or impacts on other areas of the codebase: backends are imported from TFE which we can't change. We discussed in the issue that there could be some incompatibility, depending how backends handle `-target` (as we can't force them to support `-exclude`). This doesn't initially look like it will be a problem, but may be worth noting.
```
Comment (RLRabinowitz) [inline rfc/20240611-ExcludeFlag.md:47]: This technically means that remote backends or clouds would not get the `-exclude` flag arguments in remote runs triggered from a local machine

**Review by RLRabinowitz — COMMENTED** [review]
Thank you very much for your contribution and involvement @JamesBalazs  🙏🏻 
Much appreciated

**Comment by cam72cam** [comment]
@JamesBalazs it looks like you missed the DCO Signoff on your latest commit.  You can amend that commit with the `-s` flag.

**Comment by RLRabinowitz** [comment]
Hey @JamesBalazs , thank you very much for your work with the RFC and on the issue!

I've created an RFC detailing how an `-exclude` flag should behave in https://github.com/opentofu/opentofu/pull/1860

Would love to hear your thoughts 🙏🏻

**Pull request closed without merging**